File: .agents/skills/agent-workspace-checkpoint/scripts/relay_crypto.mjs
#!/usr/bin/env node

import {
  createCipheriv,
  createDecipheriv,
  randomBytes,
} from "node:crypto";
import {
  appendFile,
  open,
  rm,
  stat,
} from "node:fs/promises";
import { createReadStream, createWriteStream } from "node:fs";
import { pipeline } from "node:stream/promises";

const MAGIC = Buffer.from("RELAYCP2\n", "ascii");
const TAG_BYTES = 16;
const NONCE_BYTES = 12;
const MAX_HEADER_BYTES = 16 * 1024;

function fail(message) {
  process.stderr.write(`${message}\n`);
  process.exitCode = 1;
}

async function readKey() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  const encoded = Buffer.concat(chunks).toString("utf8").trim();
  const key = Buffer.from(encoded, "base64url");
  if (key.length !== 32) throw new Error("Checkpoint encryption key must be 32 bytes.");
  return key;
}

function makeHeader(checkpointId, nonce) {
  const header = Buffer.from(
    JSON.stringify({
      formatVersion: 2,
      cipher: "AES-256-GCM",
      checkpointId,
      nonce: nonce.toString("base64url"),
    }),
    "utf8",
  );
  if (header.length > MAX_HEADER_BYTES) throw new Error("Checkpoint header is too large.");
  return header;
}

function makePrefix(header) {
  const length = Buffer.alloc(4);
  length.writeUInt32BE(header.length);
  return Buffer.concat([MAGIC, length, header]);
}

async function readExactly(handle, length, position) {
  const buffer = Buffer.alloc(length);
  const { bytesRead } = await handle.read(buffer, 0, length, position);
  if (bytesRead !== length) throw new Error("Encrypted checkpoint is truncated.");
  return buffer;
}

async function parseHeader(input) {
  const handle = await open(input, "r");
  try {
    const prefix = await readExactly(handle, MAGIC.length + 4, 0);
    if (!prefix.subarray(0, MAGIC.length).equals(MAGIC)) {
      throw new Error("File is not a Relay encrypted checkpoint.");
    }
    const headerLength = prefix.readUInt32BE(MAGIC.length);
    if (headerLength < 2 || headerLength > MAX_HEADER_BYTES) {
      throw new Error("Encrypted checkpoint header length is invalid.");
    }
    const headerBytes = await readExactly(handle, headerLength, MAGIC.length + 4);
    let header;
    try {
      header = JSON.parse(headerBytes.toString("utf8"));
    } catch {
      throw new Error("Encrypted checkpoint header is invalid.");
    }
    if (
      header?.formatVersion !== 2 ||
      header?.cipher !== "AES-256-GCM" ||
      typeof header?.checkpointId !== "string" ||
      !/^cp_[a-z0-9_-]{6,80}$/i.test(header.checkpointId)
    ) {
      throw new Error("Encrypted checkpoint header is unsupported.");
    }
    const nonce = Buffer.from(String(header.nonce ?? ""), "base64url");
    if (nonce.length !== NONCE_BYTES) {
      throw new Error("Encrypted checkpoint nonce is invalid.");
    }
    return {
      header,
      headerBytes,
      nonce,
      bodyStart: MAGIC.length + 4 + headerLength,
    };
  } finally {
    await handle.close();
  }
}

async function encrypt(input, output, checkpointId) {
  if (!/^cp_[a-z0-9_-]{6,80}$/i.test(checkpointId)) {
    throw new Error("Checkpoint ID is invalid.");
  }
  const key = await readKey();
  const nonce = randomBytes(NONCE_BYTES);
  const header = makeHeader(checkpointId, nonce);
  const cipher = createCipheriv("aes-256-gcm", key, nonce, { authTagLength: TAG_BYTES });
  cipher.setAAD(header);

  const handle = await open(output, "wx", 0o600);
  try {
    await handle.write(makePrefix(header));
  } finally {
    await handle.close();
  }

  try {
    await pipeline(
      createReadStream(input),
      cipher,
      createWriteStream(output, { flags: "a", mode: 0o600 }),
    );
    await appendFile(output, cipher.getAuthTag());
  } catch (error) {
    await rm(output, { force: true });
    throw error;
  }
}

async function decrypt(input, output) {
  const key = await readKey();
  const parsed = await parseHeader(input);
  const inputStat = await stat(input);
  const bodyEnd = inputStat.size - TAG_BYTES - 1;
  if (bodyEnd < parsed.bodyStart) throw new Error("Encrypted checkpoint has no ciphertext.");

  const handle = await open(input, "r");
  let tag;
  try {
    tag = await readExactly(handle, TAG_BYTES, inputStat.size - TAG_BYTES);
  } finally {
    await handle.close();
  }

  const decipher = createDecipheriv("aes-256-gcm", key, parsed.nonce, {
    authTagLength: TAG_BYTES,
  });
  decipher.setAAD(parsed.headerBytes);
  decipher.setAuthTag(tag);

  try {
    await pipeline(
      createReadStream(input, { start: parsed.bodyStart, end: bodyEnd }),
      decipher,
      createWriteStream(output, { flags: "wx", mode: 0o600 }),
    );
  } catch {
    await rm(output, { force: true });
    throw new Error("Checkpoint authentication failed. The key is wrong or the file was changed.");
  }
}

async function main() {
  const [action, input, output, checkpointId] = process.argv.slice(2);
  if (action === "encrypt" && input && output && checkpointId) {
    await encrypt(input, output, checkpointId);
    return;
  }
  if (action === "decrypt" && input && output) {
    await decrypt(input, output);
    return;
  }
  throw new Error(
    "Usage: relay_crypto.mjs encrypt INPUT OUTPUT CHECKPOINT_ID | decrypt INPUT OUTPUT",
  );
}

main().catch((error) => {
  fail(error instanceof Error ? error.message : "Checkpoint cryptography failed.");
});

File: .agents/skills/agent-workspace-checkpoint/scripts/relay_crypto.py
"""Authenticated, streaming Relay checkpoint encryption through Node.js crypto."""

from __future__ import annotations

import base64
import getpass
import hmac
import json
import re
import secrets
import shutil
import struct
import subprocess
from pathlib import Path

MAGIC = b"RELAYCP2\n"
MAX_HEADER_BYTES = 16 * 1024
HELPER = Path(__file__).with_name("relay_crypto.mjs")
MIN_PASSPHRASE_CHARACTERS = 8
KDF_NAME = "scrypt"
KDF_SALT_BYTES = 16
KDF_N = 131_072
KDF_R = 8
KDF_P = 1
CHECKPOINT_ID_PATTERN = re.compile(r"^cp_[A-Za-z0-9_-]{6,80}$")


class RelayCryptoError(RuntimeError):
    pass


def encode_secret(secret: bytes) -> str:
    if not secret:
        raise RelayCryptoError("Checkpoint encryption key cannot be empty")
    return base64.urlsafe_b64encode(secret).rstrip(b"=").decode("ascii")


def validate_checkpoint_passphrase(value: str) -> bytes:
    if len(value) < MIN_PASSPHRASE_CHARACTERS:
        raise RelayCryptoError(
            "Checkpoint passphrase must be at least 8 characters"
        )
    try:
        return value.encode("utf-8")
    except UnicodeEncodeError as error:
        raise RelayCryptoError(
            "Checkpoint passphrase contains invalid Unicode"
        ) from error


def read_checkpoint_passphrase(prompt: str) -> bytes:
    return validate_checkpoint_passphrase(getpass.getpass(prompt))


def prompt_checkpoint_secret() -> bytes:
    return read_checkpoint_passphrase(
        "Checkpoint passphrase or recovery key (minimum 8 characters): "
    )


def prompt_new_checkpoint_passphrase() -> bytes:
    passphrase = read_checkpoint_passphrase(
        "Choose checkpoint passphrase (minimum 8 characters): "
    )
    confirmation = read_checkpoint_passphrase("Confirm checkpoint passphrase: ")
    if not hmac.compare_digest(passphrase, confirmation):
        raise RelayCryptoError("Checkpoint passphrases do not match")
    return passphrase


def generate_checkpoint_recovery_key() -> bytes:
    return encode_secret(secrets.token_bytes(32)).encode("ascii")


def is_encrypted_checkpoint(path: Path) -> bool:
    try:
        with path.open("rb") as handle:
            return handle.read(len(MAGIC)) == MAGIC
    except OSError:
        return False


def read_encrypted_header(path: Path) -> dict[str, object]:
    try:
        with path.open("rb") as handle:
            if handle.read(len(MAGIC)) != MAGIC:
                raise RelayCryptoError("File is not a Relay encrypted checkpoint")
            length_bytes = handle.read(4)
            if len(length_bytes) != 4:
                raise RelayCryptoError("Encrypted checkpoint is truncated")
            header_length = struct.unpack(">I", length_bytes)[0]
            if header_length < 2 or header_length > MAX_HEADER_BYTES:
                raise RelayCryptoError("Encrypted checkpoint header length is invalid")
            header_bytes = handle.read(header_length)
            if len(header_bytes) != header_length:
                raise RelayCryptoError("Encrypted checkpoint is truncated")
        header = json.loads(header_bytes)
    except (OSError, json.JSONDecodeError, UnicodeDecodeError) as error:
        if isinstance(error, RelayCryptoError):
            raise
        raise RelayCryptoError("Encrypted checkpoint header is invalid") from error
    if (
        not isinstance(header, dict)
        or header.get("formatVersion") != 2
        or header.get("cipher") != "AES-256-GCM"
        or not isinstance(header.get("checkpointId"), str)
    ):
        raise RelayCryptoError("Encrypted checkpoint format is unsupported")
    kdf = header.get("kdf")
    if kdf is not None and (
        not isinstance(kdf, dict)
        or set(kdf) != {"name", "salt", "N", "r", "p"}
        or kdf.get("name") != KDF_NAME
        or not isinstance(kdf.get("salt"), str)
        or not re.fullmatch(r"[A-Za-z0-9_-]{22}", kdf["salt"])
        or kdf.get("N") != KDF_N
        or kdf.get("r") != KDF_R
        or kdf.get("p") != KDF_P
    ):
        raise RelayCryptoError("Encrypted checkpoint key derivation is unsupported")
    return header


def encrypt_checkpoint(
    plaintext: Path,
    encrypted: Path,
    checkpoint_id: str,
    key: bytes,
) -> None:
    _run_helper("encrypt", plaintext, encrypted, checkpoint_id, key=key)


def decrypt_checkpoint(
    encrypted: Path,
    plaintext: Path,
    key: bytes,
    expected_checkpoint_id: str | None = None,
) -> dict[str, object]:
    header = read_encrypted_header(encrypted)
    if (
        expected_checkpoint_id
        and header.get("checkpointId") != expected_checkpoint_id
    ):
        raise RelayCryptoError("Encrypted checkpoint ID does not match the request")
    _run_helper("decrypt", encrypted, plaintext, key=key)
    return header


def _run_helper(
    action: str,
    input_path: Path,
    output_path: Path,
    checkpoint_id: str | None = None,
    *,
    key: bytes,
) -> None:
    node = shutil.which("node")
    if not node:
        raise RelayCryptoError(
            "Node.js 22 or newer is required for checkpoint encryption"
        )
    command = [node, str(HELPER), action, str(input_path), str(output_path)]
    if checkpoint_id:
        command.append(checkpoint_id)
    result = subprocess.run(
        command,
        input=encode_secret(key) + "\n",
        capture_output=True,
        text=True,
        check=False,
    )
    if result.returncode:
        detail = result.stderr.strip() or "Checkpoint cryptography failed"
        raise RelayCryptoError(detail)
